components/proxy/src/main/java/com/hotels/styx/routing/handlers/HostProxy.java

```diff
@@ -0,0 +1,244 @@
+import com.hotels.styx.api.extension.service.TlsSettings;
+import com.hotels.styx.client.Connection;
+import com.hotels.styx.client.OriginStatsFactory;
+import com.hotels.styx.client.StyxHostHttpClient;
+import com.hotels.styx.client.applications.metrics.OriginMetrics;
+import com.hotels.styx.client.connectionpool.ConnectionPool;
+import com.hotels.styx.client.connectionpool.ExpiringConnectionFactory;
+import com.hotels.styx.client.connectionpool.SimpleConnectionPoolFactory;
+import com.hotels.styx.client.netty.connectionpool.NettyConnectionFactory;
+import com.hotels.styx.config.schema.Schema;
+import com.hotels.styx.infrastructure.configuration.yaml.JsonNodeConfig;
+import com.hotels.styx.routing.RoutingObject;
+import com.hotels.styx.routing.config.HttpHandlerFactory;
+import com.hotels.styx.routing.config.RoutingObjectDefinition;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import static com.hotels.styx.api.extension.Origin.newOriginBuilder;
+import static com.hotels.styx.api.extension.service.ConnectionPoolSettings.defaultConnectionPoolSettings;
+import static com.hotels.styx.client.HttpRequestOperationFactory.Builder.httpRequestOperationFactoryBuilder;
+import static com.hotels.styx.config.schema.SchemaDsl.atLeastOne;
+import static com.hotels.styx.config.schema.SchemaDsl.bool;
+import static com.hotels.styx.config.schema.SchemaDsl.field;
+import static com.hotels.styx.config.schema.SchemaDsl.integer;
+import static com.hotels.styx.config.schema.SchemaDsl.list;
+import static com.hotels.styx.config.schema.SchemaDsl.object;
+import static com.hotels.styx.config.schema.SchemaDsl.optional;
+import static com.hotels.styx.config.schema.SchemaDsl.string;
+import static com.hotels.styx.routing.config.RoutingSupport.missingAttributeError;
+import static java.lang.String.format;
+import static java.lang.String.join;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+/**
+ * A routing object that proxies all incoming traffic to a remote host.
+ */
+public class HostProxy implements RoutingObject {
+    public static final Schema.FieldType SCHEMA = object(
+            field("host", string()),
+            optional("tlsSettings", object(
+                    optional("trustAllCerts", bool()),
+                    optional("sslProvider", string()),
+                    optional("trustStorePath", string()),
+                    optional("trustStorePassword", string()),
+                    optional("protocols", list(string())),
+                    optional("cipherSuites", list(string())),
+                    optional("additionalCerts", list(object(
+                            field("alias", string()),
+                            field("certificatePath", string())
+                    ))),
+                    atLeastOne("trustAllCerts",
```

Comment:
is this validation correct? It seems that "protocols" and "cipherSuites" are not related with trustAllCerts/trustStorePath/additionalCerts and, of course, trustStorePassword could not be the "only one" here.

Reply:
Hi @dvlato could you clarify this please?

>  It seems that "protocols" and "cipherSuites" are not related with trustAllCerts/trustStorePath/additionalCerts and,

This schema models the existing configuration format using the rudimentary validator that is available at the moment.

For example, there is no way to tell the validator to fail the configuration if `trustStorePassword` appears without `trustStorePath`.

Reply:
As discussed offline this can be ignored.